src/katgpucbf/xbgpu/engine.py

```diff
@@ -282,23 +283,22 @@ def __init__(
         super().__init__(katcp_host, katcp_port)
         self._cancel_tasks: list[asyncio.Task] = []  # Tasks that need to be cancelled on shutdown
 
+        # B-engine doesn't work yet
+        assert len(outputs) == 1
+        assert isinstance(outputs[0], XengineOutput)
+
         if sample_bits != 8:
             raise ValueError("sample_bits must equal 8 - no other values supported at the moment.")
 
-        if channel_offset_value % n_channels_per_stream != 0:
-            raise ValueError("channel_offset must be an integer multiple of n_channels_per_stream")
+        if channel_offset_value % outputs[0].channels_per_substream != 0:
+            raise ValueError("channel_offset must be an integer multiple of channels_per_substream")
 
         # Array configuration parameters
         self.adc_sample_rate_hz = adc_sample_rate_hz
-        self.send_rate_factor = send_rate_factor
         self.heap_accumulation_threshold = heap_accumulation_threshold
         self.time_converter = TimeConverter(sync_epoch, adc_sample_rate_hz)
-        self.n_ants = n_ants
-        self.n_channels_total = n_channels_total
-        self.n_channels_per_stream = n_channels_per_stream
-        self.n_spectra_per_heap = n_spectra_per_heap
+        self.n_ants = outputs[0].antennas  # Still needed in sender_loop
```

Comment:
There is a bit of a smell in the number of antennas being obtained from an Output, given that it's determined by the input. I'm not sure it belongs in the Output class at all, although I can see it may be convenient to have there for computing things like the `baselines` property. I'd suggest removing it from there for now and making everything else work (to ensure that the input side of things doesn't depend on having a X output) then think about adding it back.

Reply:
I'm with you. As I mentioned in [an earlier reply](https://github.com/ska-sa/katgpucbf/pull/568#discussion_r1197425815), you're entirely correct Re:
> [...] I can see it may be convenient to have there for computing things like the `baselines` property.

Shall move forward with your suggestion.